### PR TITLE
Add support for wasi-http in the wasm binding.

### PR DESCRIPTION
**Pull request by brendandburns**
# Description

Add support for `wasi-http` client calls from the WASM binding. This enables developers of WASM bindings to make calls into Dapr (e.g. storage) or any other URL.

I will add documentation once there is agreement that this PR is ready to merge, but prior to merging it.

## Issue reference

See discussion in #2990 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [X] Code compiles correctly
* [X] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by ItalyPaleAle** [comment]
Thanks @brendandburns !

CC: @codefromthecrypt @Taction Can you please take a look

**Comment by codefromthecrypt** [comment]
Message ID: ***@***.***>notes on the specifics. i wont repeat the points on the general.

wasip1 and wasi-http are not mutually exclusive, so the imports part needs to switch to bitflags

That this works knowing wasip1 makes me think that it is not using wazero's wasip1 impl (rather stealthrocket which is different syscall  code based on x/sys iirc). If correct, this means strictmode will be totally ignored and real everything is used. If i am correct, this would become obvious if bitflags were used as after wasip1 loads stealthrocket will also try to load it and crash. In general there should be an "instantiate" function (and bitflag) for each wasm module wasi-http code imports. iirc there are a few as component model is unlike wasip1 which is monolithic. Brendan should know what these are and help explain. in any case hidden module deps and behavior should at least be comments at the call site here like.. "// this uses an alternate wasip1 implentation that doesnt support sandboxing. wasi-http itself imports the following three modules: x, y, z. See https://path/to/wasi-http-spec

The binary checked in is huge (5.6m). I suspect it wasmt built with the makefile so please configure the makefile and build it with that from source here. This can give a better sense.


Limitation should be noted that this binding config will not work with Go 1.21, documentation should be careful to not mislead anyone into thinking that.

technical point: this is designed tinygo can work but not regular go. unlike http-wasm, wasi-http shares memory responsibilty with the host (via exported malloc like functions). These work in tinygo because it can export functions besides main (_start) and also it can directly expose a malloc function (maybe it is remalloc)

Go 1.21 does not, so when out http-wasm client is out soon, only that will work with both tinygo and go. We designed http-wasm speciially for the guest handling all memory concerns. It was not exactly for this reason, but other issues when guest and host share memory which were problems back then in proxy-wasm. A nice side effect is that the client will not need to export functions for the same reason and that means it will work with Go 1.21 GOOS=wasip1 which will be released publicly in a couple weeks.

**Comment by codefromthecrypt** [comment]
Another important point and I don't know if it is relevant. wasi-go doesnt work on windows. 

If dapr needs to build on windows you should double-check here either via tests or via 'GOOS=windows go build ...'  It is possible that through my suggestions wasi-http could avoid windows incompatible parts and use wazero's imports for those instead. Disregard if there arent any windows developers.

**Comment by brendandburns** [comment]
@codefromthecrypt @ItalyPaleAle 

regarding checking in the Makefile, the main issue is for this to work we need to add a dependency on the `github.com/dev-wasm/dev-wasm-go/http/client` package.

Sadly go doesn't allow for dev dependencies or sub-folder dependencies, which means that this would be a top-level dependency for the dapr components package.

I could also version the client into the `testdata` folder if we wanted to, but that's also a little hacky since it will need to be kept in sync.

Let me know what the preference is. I'm happy to add a `Makefile` but I don't think that it will work out of the box without adding some dependencies to the `go.mod` file.

Regarding the `wasi1` I haven't made any changes, this is already not mutually exclusive, and it is the `wazero` wasi implementation. If you look at the code, I set a boolean in a `map` depending on what is required and I load in those bits as necessary.

I believe that while `wasi-go` may not work with Windows (I haven't validated this) the `wasi-http` bits (and that's all that's being added) should work correctly with windows. I will validate when I get onto a windows box later today.

**Comment by codefromthecrypt** [comment]
Thanks for thinking things through I'll reply to a couple things which I've specific opinions on and may know more by default than others.

I'm giving feedback for you and others, but remain of the opinion this should not be placed into the same component as the default wasm binding.

### On guest..

The very original version of middleware had an opaque wasm file which meant we couldn't know if as a unit anything worked or not. So, we stopped that and have guests at least rebuildable here. This makes it obvious what the ABI requirements are, as if you rebuild something it breaks, but you can see the commit it last worked on.

The way out is to add a directory here with a go.mod in it https://github.com/dapr/components-contrib/tree/master/bindings/wasm/testdata, you may need to update the Makefile. That it uses the SDK you mention is a reality of the whole product, so hiding that I think isn't a goal if this is supposed to be supported.

This doesn't create a runtime dependency because the go.mod file breaks the tree of components-contrib. It is similar to the middleware which has a separate go.mod for the guest side of the handler https://github.com/dapr/components-contrib/blob/master/middleware/http/wasm/example/go.mod

OTOH, as this is a wasi (black box subprocess model) it isn't required by me personally to use go at all. Just I think for maintainers here probably using tinygo is better than Go. My 2p

### On windows..

The job that builds windows doesn't run on PR, so it is hard to detect until post merge. Actually you can see that fail here https://github.com/dapr/components-contrib/actions/runs/5627721004/job/15250692975 In wasi-go I think they can make a way to ensure this part compiles as you understand it.

### On bitflag vs map of bools..

I agree you are solving in a different way than bitflag. If you don't feel strongly about it, I'd recommend using a style more similar to what's going on inside the handler code. OTOH I don't have a strong feeling on this, I'll leave it up to you and the maintainers.

https://github.com/http-wasm/http-wasm-host-go/blob/main/handler/middleware.go#L764-L782
https://github.com/http-wasm/http-wasm-host-go/blob/main/handler/middleware.go#L94-L108

What made me curious was both that there's overlap in wasi-go and also multiple modules nested inside wasi-http. I did end up looking just now, to verify.

Everywhere else in the wazero ecosystem modules are configured one at a time, and this one configures three https://github.com/stealthrocket/wasi-go/blob/main/imports/wasi_http/http.go#L12-L23

I think that's worth a comment... what's going on. Even better if you can be explicit about all three and only instantiate the ones used.

**Comment by codefromthecrypt** [comment]
ps I didn't mean to imply the current windows build fail has anything to do with this PR, just noting in this repo you can't tell before merge if it will break unless you run manually.

**Comment by codefromthecrypt** [comment]
Overall point on sandboxing, more for @ItalyPaleAle. So I personally wrote a host side http impl for GOOS=js (predecessor to wasip1). I think that in Go, the best way to handle configuration is via giving a roundtripper config (that's how I did that before). This allows you 100pct control over how things work in a way idiomatic and easy to understand. OTOH this may not fit the underlying impl of this change. Anyway that's what the http-wasm option will do, and mentioning it just in case other ABI would like to provide that style also.

**Comment by codefromthecrypt** [comment]
@brendandburns For this specific change, to control I think easiest way is to replace code like this with a roundTripper that can be provided externally https://github.com/stealthrocket/wasi-go/blob/main/imports/wasi_http/types/request.go#L67C24-L67C24  The sandbox one would return some well-known error so that there's no risk of escaping, and in any case the default may choose to limit to localhost. Once this config is externalized dapr can make whatever decision it wants to.

**Comment by codefromthecrypt** [comment]
I'm not going to be around for at least a week or two, so please don't block on my comments for anything. I've said everything I can think of. Good luck whichever path you take.

File: bindings/wasm/output.go
Comment (brendandburns) [reply bindings/wasm/output.go:91]: Done.

**Comment by brendandburns** [comment]
Added a Makefile that shows how to build this with tinygo, added go.mod and go.sum files (thanks for the pointer).

In case this wasn't clear, the goal of this is to target all languages that can be compiled to WASM, not just Go. I think that polyglot supports adds value. That's a central benefit of the generated bindings, it's easy to target multiple different guests.

Added the strict mode support that @ItalyPaleAle was asking for.

For the multiple modules, this is because of the way that the `wit-bindgen` generator generated the bindings. The `streams` module is a generic part of the `wasi` specification that will be used in `http` but also in other bindings.

**Comment by ItalyPaleAle** [comment]
Thanks for this PR @brendanburns !

It seems that the CI isn't fully happy yet. Can you please run `make modtidy-all`?

Aside from this, LGTM :)

**Review by ItalyPaleAle — APPROVED** [review]

**Comment by brendandburns** [comment]
Just a quick response to the Windows comment above, I have tested and validated that this works correctly (at leasts tests pass) on Windows in addition to Linux.

**Pull request merged**
